cou/steps/plan.py

```diff
@@ -101,13 +103,14 @@ async def create_upgrade_group(
     :type filter_function: Callable[[OpenStackApplication], bool]
     :raises Exception: When cannot generate upgrade plan.
     :return: Upgrade group.
-    :rtype: UpgradeStep
+    :rtype: Optional[UpgradeStep]
     """
     group_upgrade_plan = UpgradeStep(description=description, parallel=False)
     for app in filter(filter_function, apps):
         try:
-            app_upgrade_plan = app.generate_upgrade_plan(target)
-            group_upgrade_plan.add_step(app_upgrade_plan)
+            app_upgrade_plan: Optional[UpgradeStep] = app.generate_upgrade_plan(target)
+            if app_upgrade_plan:
+                group_upgrade_plan.add_step(app_upgrade_plan)
```

Comment:
If we define empty property step we can do `if.app_upgrade_plan.empty`, which will be similar to what we have here. But what if we simply not show empty step in string representation and in execution we can skip any empty step. Like this we will have information about step, but we will know that's empty.